modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -806,4 +866,230 @@ private static Throwable unwrapDistributionZoneExceptionRecursively(Throwable e,
+
+        private final Set<String> nodesToRemove;
+
+        ZoneState() {
+            this.lockDeltas = new ReentrantLock();
+            this.lockForTimers = new ReentrantLock();
+            this.nodesToAdd = new HashSet<>();
+            this.nodesToRemove = new HashSet<>();
+        }
+
+        private void rescheduleScaleUp(long delay, Runnable runnable) {
+            if (scaleUp == null) {
+                scaleUp = new Timer(runnable);
+            }
+
+            scaleUp.reschedule(delay);
+        }
+
+        private void rescheduleOngoingScaleUp(long delay) {
+            if (scaleUp != null) {
+                scaleUp.rescheduleOngoingTask(delay);
+            }
+        }
+
+        private void rescheduleScaleDown(long delay, Runnable runnable) {
+            if (scaleDown == null) {
+                scaleDown = new Timer(runnable);
+            }
+
+            scaleDown.reschedule(delay);
+        }
+
+        public ReentrantLock lock() {
+            return lockDeltas;
+        }
+
+        public ReentrantLock lockForTimers() {
+            return lockForTimers;
+        }
+
+        public Set<String> nodesToAdd() {
+            return nodesToAdd;
+        }
+
+        public Set<String> nodesToRemove() {
+            return nodesToRemove;
+        }
+
+        public void nodesToAdd(Set<String> nodes) {
+            lockDeltas.lock();
+
+            try {
+                nodesToAdd.addAll(nodes);
+            } finally {
+                lockDeltas.unlock();
+            }
+        }
+
+        public void nodesToRemove(Set<String> nodes) {
+            lockDeltas.lock();
+
+            try {
+                nodesToRemove.addAll(nodes);
+            } finally {
+                lockDeltas.unlock();
+            }
+        }
+    }
+
+    private class Timer implements Runnable {
+        private final Runnable runnable;
+
+        ScheduledFuture<?> task;
+
+        Timer(Runnable runnable) {
+            this.runnable = runnable;
+        }
+
+        /** {@inheritDoc} */
+        @Override public void run() {
+            runnable.run();
+        }
+
+        private void reschedule(long delay) {
+            if (task != null) {
+                task.cancel(false);
+            }
+
+            task = executor.schedule(this, delay, TimeUnit.MILLISECONDS);
+        }
+
+        private void rescheduleOngoingTask(long delay) {
+            if (task != null) {
+                if (task.cancel(false)) {
+                    task = executor.schedule(this, delay, TimeUnit.MILLISECONDS);
+                }
+            }
+        }
+    }
+
+    private void initTriggerKeysAndDataNodes(int zoneId) {
+        If iifScaleUp = If.iif(
+                notExists(zoneScaleUpChangeTriggerKey(zoneId)),
+                ops(put(zoneScaleUpChangeTriggerKey(zoneId), ByteUtils.longToBytes(0L))).yield(true),
+                ops().yield(false)
+        );
+
+        metaStorageManager.invoke(iifScaleUp);
```

Comment:
Here and in updateMetaStorage... you are ignoring invoke result. It's probably not a good idea ;)